README
Easy installation without kernel patch
--------------------------------------

1) Run 'make' and 'make install'. It builds and installs the module
   for the running kernel version as reported by 'uname -r'.
   If you want to build for another kernel, edit the Makefile.

2) Either run 'modprobe fusion' manually or add "fusion" to "/etc/modules".

3) Create the fusion device(s) if not using devfs or udev:

	mkdir /dev/fusion
	mknod /dev/fusion/0 c 250 0

	...if you need more than one session

	mknod /dev/fusion/1 c 250 1
	mknod /dev/fusion/2 c 250 2

	...and so on (currently limited to eight sessions)

4) Add udev rules to /etc/udev/rules.d/40-fusion.rules if using udev:

	KERNEL=="fusion[0-9]*", NAME="fusion/%n", GROUP="video", MODE="0660"

	...customize to suit your needs


ChangeLog
commit b7a00091b78943cd50ca16aff502d6eee9765a64
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Mon Jul 15 11:55:25 2013 +0200

    debian: fix header location in rules script

commit c89972a75160222cc209349d024883c697af9d46
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Mon Jul 15 11:34:11 2013 +0200

    update changelog

commit a46139ef2570de106799b235203eb22185239180
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Mon Jul 15 11:33:47 2013 +0200

    change version to 9.0.2

commit 5858093828e6f9e8b9a0fb7df008f34d81e80e20
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Mon Jul 15 11:29:35 2013 +0200

    do not access members of proc_dir_entry anymore (for linux 3.10)

commit c8948946261f2a8e6cbf12b1b33791a29fcdb7a3
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Fri Jul 12 11:57:16 2013 +0200

    fix one for 3.10, move PDE_DATA compat macro from fusiondev.h to types.h
    
    still need a replacement for proc_dir_entry::name

commit 7b276feeeb847d9c6ab8f73a15e9eaa2daafb088
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Thu Jul 11 22:13:46 2013 +0200

    fusion: support linux 3.10 (incomplete)
    
    TODO:
    - do the same for one
    - find a replacement for proc_dir_entry::name

commit 0f3669626cb4307d92839184a94b60ccec0c3863
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Fri Jul 5 17:24:49 2013 +0200

    update changelog

commit c186c708f39feb68b00c68953b1f5ae22b0344ec
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Fri Jul 5 17:24:05 2013 +0200

    debian: 9.0.1

commit a76a43b33b2d4a83af4ecbe1243dbbbdf0decee9
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Sat Jun 15 22:06:03 2013 +0200

    proc: workaround dead lock issue when shutting down fusion device while opening proc file

commit 8eff11a0a0621307c24e1644cbf8cabb907d5787
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Sat Jun 15 22:05:33 2013 +0200

    API: update comment of quota limit

commit d864cf9d80ebcf4c2d5222b0d18853f4f360b347
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Sat Jun 15 22:04:48 2013 +0200

    call: fix warning

commit 65ff77e8993d0bb98a99d46e6a4bc7bb6fa8e5cd
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Wed Jun 5 15:26:14 2013 +0200

    fusion/shmpool: fix memory corruption bug with multiple fusion worlds on 32-bit machines

commit 2c8f51d108da9c97d111d4631f614f099738710d
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Sun Apr 21 10:50:41 2013 +0200

    call: Some quota fixes and debug.
    
    Signed-off-by: Denis Oliver Kropp <dok@directfb.org>

commit 670da5f481d124ee4bb5758de83ccf1997fe7a22
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Fri Apr 19 19:34:40 2013 +0200

    call: Wake waiter in remove_execution.

commit 4180a374fc061d453eb77a54d314e7245d7f7fd5
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Fri Apr 19 19:33:08 2013 +0200

    call: Don't wake waiters on execution in free_all_executions.

commit 65267d45552d8d4c422ace39cde02b34c48c304a
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Fri Apr 19 19:30:11 2013 +0200

    call: Don't need waiting for messages in call_destroy, waiting for executions afterwards anyways.

commit 68b45d71cb82c182b8347de30bc788b147475cb4
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Sat Mar 30 10:31:02 2013 +0100

    call: Print call id and serial in execute functions.

commit 59d160f3f332f25246e74f77cc93c4c070327479
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Sat Mar 30 10:28:33 2013 +0100

    call: Fix another resume issue by not removing the execution on return, but after the result is received on caller side.

commit 7644d9b148d72a5f872338f7e7f2d6ee774744be
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Fri Mar 29 07:47:35 2013 +0100

    call: Fix FCEF_RESUMABLE calls (EINTR) with FCEF_FOLLOW flag (buffered calls being flushed).
    
    This fixes EIDRM being received after a signal with restart of multiple calls in fusion_world_flush_calls().

commit 29aae5f09a594ab3b26a6c012997ea54004d7338
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Fri Mar 29 07:29:09 2013 +0100

    call: print error when execution has not been found on resume

commit 10e49d9f689d96842d6232d8574e7bd813e9027c
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Thu Mar 14 01:10:38 2013 +0100

    FIXME: use entry lookup via hash table, or add reactor pointer to dispatch if possible

commit a91ffe0ec61b76a8e448d92cbe789d5bcbe992a7
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Thu Mar 14 01:09:44 2013 +0100

    fusionee: Fix fusionee_sync() by waiting for the dispatcher to block on a read.

commit c781ec23a4be64cdaea61056f4d9a4340602b0d0
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Thu Mar 14 01:08:04 2013 +0100

    call: Fix dead lock with user space refs by allowing packet over-allocation (non-blocking) for calls from dispatcher thread itself.

commit 83baf1250027b213987e60cdd46b9d10d37435f0
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Thu Mar 14 01:07:25 2013 +0100

    fusiondev: Increase number of minors from 8 to 32 :)

commit 74665374b8d2cf96eb80b4872dbeff044c4283d0
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Thu Mar 14 01:05:54 2013 +0100

    shm: Add FUSION_SHM_PER_WORLD_SPACE to statically map the different worlds in a bigger address range, enabled on 64bit machines.

commit 42b233dad44abb2408a22867657f11bc08ce6d55
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Fri Jan 18 21:21:16 2013 +0100

    debian: install one.h

commit 31a489822df20db189f407e663102bc115eac9c8
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Fri Jan 18 15:12:03 2013 +0100

    update changelog

commit a3dc69d85f5cd88a1a700bf510db43e83b18c842
Author: Andreas Shimokawa <andi@directfb.org>
Date:   Tue Jan 15 11:33:16 2013 +0100

    debian: update package version

commit 06c65ec780713ed1d836cd0d2b21672b0cc2bc59
Author: Denis Oliver Kropp <dok@directfb.org>
Date:   Fri Jan 11 13:12:03 2013 +0100


one/onedev.c
/*
   (c) Copyright 2011  Denis Oliver Kropp <dok@directfb.org>

   All rights reserved.

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version
   2 of the License, or (at your option) any later version.
   
   --
   
   Device is the entry from user space into One IPC implementation in kernel
   
   Each new open() from a process results in creation of a One Application
*/

//#define ONE_ENABLE_DEBUG

#include <linux/version.h>
#include <linux/module.h>
#ifdef HAVE_LINUX_CONFIG_H
#include <linux/config.h>
#endif
#include <linux/types.h>
#include <linux/kernel.h>
#include <linux/fs.h>
#include <linux/slab.h>
#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 18)
#include <linux/devfs_fs_kernel.h>
#endif
#if LINUX_VERSION_CODE > KERNEL_VERSION(2, 6, 19)
#include <linux/page-flags.h>
#include <linux/mm.h>
#endif
#include <asm/pgtable.h>
#include <linux/mm.h>

#include <linux/vmalloc.h>
#include <linux/mman.h>
#include <linux/proc_fs.h>
#include <linux/seq_file.h>
#include <linux/poll.h>
#include <linux/sched.h>
#include <linux/init.h>
#include <asm/io.h>
#include <asm/uaccess.h>

#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
#include <linux/device.h>
#endif

#include <linux/one.h>

#include "app.h"
#include "onedev.h"
#include "queue.h"

#ifndef ONE_MAJOR
#define ONE_MAJOR 250
#endif

MODULE_LICENSE("GPL");
MODULE_AUTHOR("Denis Oliver Kropp <dok@directfb.org>");

struct proc_dir_entry *proc_one_dir;

static int one_major = ONE_MAJOR;

#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 0)
static devfs_handle_t devfs_handles[NUM_MINORS];
static inline unsigned iminor(struct inode *inode)
{
     return MINOR(inode->i_rdev);
}
#endif

#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
static struct class *one_class;
#else
static struct class_simple *one_class;
#endif
#endif



OneDev                 one_devs[NUM_MINORS];

OneCore               *one_core;
struct proc_dir_entry *one_proc_dir[NUM_MINORS]   = { 0 };
unsigned int           one_local_refs[NUM_MINORS] = { 0 };

/******************************************************************************/
static int
onedev_apps_proc_show(struct seq_file *m, void *v)
{
     OneDev *dev = m->private;
     OneApp *app;

     seq_printf(m, "  pid   OneID       status         queues\n");

     direct_list_foreach (app, dev->apps) {
          OneAppTargetData *recv_data;

          seq_printf(m, "(%5d) 0x%08x  %s  ", current->pid,
                        app->one_id, app->recv_data ? "RECEIVING" : "not receiving");

          direct_list_foreach (recv_data, app->recv_data)
               seq_printf(m, " 0x%08x", recv_data->queue_id);

     }

     return 0;
}

static int
onedev_stat_proc_show(struct seq_file *m, void *v)
{
     OneDev *dev = m->private;

     if ((dev->api.major != 0) || (dev->api.minor != 0))
          seq_printf(m, "One API:%d.%d\n", dev->api.major,
                  dev->api.minor);

     seq_printf(m, "    attach     detach   dispatch\n");

     seq_printf(m, "%10d %10d %10d\n",
                   dev->stat.queue_attach,
                   dev->stat.queue_detach,
                   dev->stat.queue_dispatch);

     return 0;
}

/******************************************************************************/

static int onedev_apps_proc_open(struct inode *inode, struct file *file)
{
     return single_open(file, onedev_apps_proc_show, PDE_DATA(inode));
}

static int onedev_stat_proc_open(struct inode *inode, struct file *file)
{
     return single_open(file, onedev_stat_proc_show, PDE_DATA(inode));
}
 
static const struct file_operations onedev_apps_proc_fops = {
     .open    = onedev_apps_proc_open,
     .read    = seq_read,
     .llseek  = seq_lseek,
     .release = seq_release,
};

static const struct file_operations onedev_stat_proc_fops = {
     .open    = onedev_stat_proc_open,
     .read    = seq_read,
     .llseek  = seq_lseek,
     .release = seq_release,
};


static int onedev_init(OneDev * dev)
{
     int ret;

     ret = one_queue_init(dev);
     if (ret)
          goto error_queue;

     proc_create_data("apps", 0, one_proc_dir[dev->index],
                       &onedev_apps_proc_fops, dev);

     proc_create_data("stat", 0, one_proc_dir[dev->index],
                       &onedev_stat_proc_fops, dev);

     return 0;

error_queue:
     return ret;
}

static void onedev_deinit(OneDev * dev)
{
     remove_proc_entry("apps", one_proc_dir[dev->index]);
     remove_proc_entry("stat", one_proc_dir[dev->index]);

     one_queue_deinit(dev);

     one_core_cleanup( one_core );
}

/******************************************************************************/

static int one_open(struct inode *inode, struct file *file)
{
     int     ret;
     OneApp *oneapp;
     int     minor = iminor(inode);
     OneDev *dev   = &one_devs[minor];
     char buf[4];

     ONE_DEBUG("one_open( file %p, f_count %ld ) <- minor %d\n", file, atomic_long_read(&file->f_count), minor);

     snprintf(buf, 4, "%d", minor);

     one_core_lock( one_core );

     ONE_DEBUG("  -> refs: %d\n", dev->refs);

     if (!dev->refs) {
          memset( dev, 0, sizeof(OneDev) );

          dev->index = minor;
     }
     else {
          if (file->f_flags & O_EXCL) {
               if (dev->one_ids) {
                    one_core_unlock( one_core );
                    return -EBUSY;
               }
          }
     }

     if (!one_local_refs[dev->index]) {
          one_proc_dir[minor] = proc_mkdir(buf, proc_one_dir);

          ret = onedev_init( dev );
          if (ret) {
               remove_proc_entry(buf, proc_one_dir);
               one_core_unlock( one_core );
               return ret;
          }
     }

     ret = OneApp_New( dev, dev->one_ids + 1, &oneapp );
     if (ret) {
          if (!one_local_refs[dev->index]) {
               onedev_deinit( dev );
               remove_proc_entry( buf, proc_one_dir );
          }

          one_core_unlock( one_core );

          return ret;
     }

     direct_list_append( &dev->apps, &oneapp->link );

     dev->one_ids++;

     one_local_refs[dev->index]++;

     dev->refs++;

     one_core_unlock( one_core );


     file->private_data = oneapp;

     return 0;
}

static int one_release(struct inode *inode, struct file *file)
{
     int     minor  = iminor(inode);
     OneApp *oneapp = file->private_data;
     OneDev *dev    = oneapp->dev;
     char buf[4];

     ONE_DEBUG( "one_release( %p, f_count %ld )\n", file, atomic_long_read(&file->f_count) );

     snprintf(buf, 4, "%d", minor);

     one_core_lock( one_core );

     direct_list_remove( &dev->apps, &oneapp->link );

     OneApp_Destroy( oneapp );

     dev->refs--;

     one_local_refs[dev->index]--;

     if (!one_local_refs[dev->index]) {
          onedev_deinit( dev );

          remove_proc_entry( buf, proc_one_dir );
     }

     one_core_unlock( one_core );

     return 0;
}

static int
lounge_ioctl(OneDev * dev, OneApp * oneapp,
             unsigned int cmd, unsigned long arg)
{
     int          ret;
//     OneEnter     enter;
     OneEntryInfo info;

     switch (_IOC_NR(cmd)) {
#if 0
          case _IOC_NR(ONE_ENTER):
               if (copy_from_user(&enter, (OneEnter *) arg, sizeof(enter)))
                    return -EFAULT;

               ret = oneapp_enter(dev, &enter, oneapp);
               if (ret)
                    return ret;

               if (copy_to_user((OneEnter *) arg, &enter, sizeof(enter)))
                    return -EFAULT;

               return 0;

          case _IOC_NR(ONE_SYNC):
               return oneapp_sync( dev, oneapp );
#endif

          case _IOC_NR(ONE_ENTRY_SET_INFO):
               if (copy_from_user
                   (&info, (OneEntryInfo *) arg, sizeof(info)))
                    return -EFAULT;

               switch (info.type) {
                    case ONE_QUEUE:
                         return one_entry_set_info(&dev->queue, &info);

                    default:
                         return -ENOSYS;
               }

          case _IOC_NR(ONE_ENTRY_GET_INFO):
               if (copy_from_user
                   (&info, (OneEntryInfo *) arg, sizeof(info)))
                    return -EFAULT;

               switch (info.type) {
                    case ONE_QUEUE:
                         ret = one_entry_get_info(&dev->queue, &info);
                         break;

                    default:
                         return -ENOSYS;
               }

               if (ret)
                    return ret;

               if (copy_to_user((OneEntryInfo *) arg, &info, sizeof(info)))
                    return -EFAULT;

               return 0;
     }

     return -ENOSYS;
}

static int
queue_ioctl( OneDev        *dev,
             OneApp        *oneapp,
             unsigned int   cmd,
             unsigned long  arg )
{
     int                     ret;
     OneQueueNew             new_;
     OneQueueDestroy         destroy;
     OneQueueAttach          attach;
     OneQueueDetach          detach;
     OneQueueDispatch        dispatch;
     OneQueueReceive         receive;
     OneQueueDispatchReceive dispatch_receive;
     OneQueueWakeUp          wakeup;

     switch (_IOC_NR(cmd)) {
          case _IOC_NR(ONE_QUEUE_NEW):
               if (copy_from_user(&new_,
                                  (OneQueueNew *) arg,
                                  sizeof(new_)))
                    return -EFAULT;

               ret = one_queue_new( oneapp, &new_ );
               if (ret)
                    return ret;

               if (copy_to_user((OneQueueNew *) arg,
                                &new_,
                                sizeof(new_)))
                    return -EFAULT;
               return 0;

          case _IOC_NR(ONE_QUEUE_DESTROY):
               if (copy_from_user(&destroy,
                                  (OneQueueDestroy *) arg,
                                  sizeof(destroy)))
                    return -EFAULT;

               return one_queue_destroy( oneapp, &destroy );

          case _IOC_NR(ONE_QUEUE_ATTACH):
               if (copy_from_user(&attach,
                                  (OneQueueAttach *) arg,
                                  sizeof(attach)))
                    return -EFAULT;

               return one_queue_attach( oneapp, &attach );

          case _IOC_NR(ONE_QUEUE_DETACH):
               if (copy_from_user(&detach,
                                  (OneQueueDetach *) arg,
                                  sizeof(detach)))
                    return -EFAULT;

               return one_queue_detach( oneapp, &detach );

          case _IOC_NR(ONE_QUEUE_DISPATCH):
               if (copy_from_user(&dispatch,
                                  (OneQueueDispatch *) arg,
                                  sizeof(dispatch)))
                    return -EFAULT;

               return one_queue_dispatch( oneapp, &dispatch );

          case _IOC_NR(ONE_QUEUE_RECEIVE):
               if (copy_from_user(&receive,
                                  (OneQueueReceive *) arg,
                                  sizeof(receive)))
                    return -EFAULT;

               ret = one_queue_receive( oneapp, &receive );

               if (copy_to_user((OneQueueReceive *) arg,
                                &receive,
                                sizeof(receive)))
                    return -EFAULT;
               return ret;

          case _IOC_NR(ONE_QUEUE_DISPATCH_RECEIVE):
               if (copy_from_user(&dispatch_receive,
                                  (OneQueueDispatchReceive *) arg,
                                  sizeof(dispatch_receive)))
                    return -EFAULT;

               ret = one_queue_dispatch_receive( oneapp, &dispatch_receive );

               if (copy_to_user((OneQueueDispatchReceive *) arg,
                                &dispatch_receive,
                                sizeof(dispatch_receive)))
                    return -EFAULT;
               return ret;

          case _IOC_NR(ONE_QUEUE_WAKEUP):
               if (copy_from_user(&wakeup,
                                  (OneQueueWakeUp *) arg,
                                  sizeof(wakeup)))
                    return -EFAULT;

               return one_queue_wakeup( oneapp, &wakeup );
     }

     return -ENOSYS;
}

#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 36)
static long
one_ioctl(struct file *file, unsigned int cmd, unsigned long arg)
#else
static int
one_ioctl(struct inode *inode, struct file *file,
             unsigned int cmd, unsigned long arg)
#endif
{
     int     ret    = -ENOSYS;
     OneApp *oneapp = file->private_data;
     OneDev *dev    = oneapp->dev;

//     ONE_DEBUG("one_ioctl (0x%08x)\n", cmd);

     one_core_lock( one_core );

     switch (_IOC_TYPE(cmd)) {
          case ONE_LOUNGE:
               ret = lounge_ioctl(dev, oneapp, cmd, arg);
               break;

          case ONE_QUEUE:
               ret = queue_ioctl(dev, oneapp, cmd, arg);
               break;
     }

     one_core_unlock( one_core );

     return ret;
}

static struct file_operations one_fops = {
     .owner = THIS_MODULE,
     .open = one_open,
     .release = one_release,
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 36)
     .unlocked_ioctl = one_ioctl,
#else
     .ioctl = one_ioctl,
#endif
};

/******************************************************************************/

#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
static int __init register_devices(void)
{
     int i;

     one_major = ONE_MAJOR;

     if (register_chrdev(one_major, "one", &one_fops)) {
          one_major = register_chrdev(0, "one", &one_fops);
          if (one_major <= 0) {
               printk (KERN_ERR "one: unable to register one device\n");
               return -EIO;
          }
          printk(KERN_ERR "one: unable to register major %d. Registered %d instead\n", ONE_MAJOR, one_major);
     }
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
     one_class = class_create(THIS_MODULE, "one");
#else
     one_class = class_simple_create(THIS_MODULE, "one");
#endif
     if (IS_ERR(one_class)) {
          unregister_chrdev(one_major, "one");
          return PTR_ERR(one_class);
     }
#endif

#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 18)
     devfs_mk_dir("one");
#endif

     for (i = 0; i < NUM_MINORS; i++) {
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 27)
          device_create(one_class,
                        NULL,
                        MKDEV(one_major, i), NULL, "one%d", i);
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 26)
          device_create(one_class,
                        NULL, MKDEV(one_major, i), "one%d", i);
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 15)
          class_device_create(one_class,
                              NULL,
                              MKDEV(one_major, i),
                              NULL, "one%d", i);
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
          class_device_create(one_class,
                              MKDEV(one_major, i),
                              NULL, "one%d", i);
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
          class_simple_device_add(one_class,
                                  MKDEV(one_major, i),
                                  NULL, "one%d", i);
#endif

#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 18)
          devfs_mk_cdev(MKDEV(one_major, i),
                        S_IFCHR | S_IRUSR | S_IWUSR, "one/%d", i);
#endif
     }

     return 0;
}
#else
static int __init register_devices(void)
{
     int i;
     char buf[16];

     if (devfs_register_chrdev(one_major, "one", &one_fops)) {
          printk(KERN_ERR "one: unable to get major %d\n",
                 one_major);
          return -EIO;
     }

     for (i = 0; i < NUM_MINORS; i++) {
          snprintf(buf, 16, "one/%d", i);

          devfs_handles[i] = devfs_register(NULL, buf, DEVFS_FL_DEFAULT,
                                            one_major, i,
                                            S_IFCHR | S_IRUSR | S_IWUSR,
                                            &one_fops, NULL);
     }

     return 0;
}
#endif



int __init one_init(void)
{
     int ret;

     printk( KERN_INFO "%s()\n", __FUNCTION__ );

     one_core_enter( &one_core );

     ret = register_devices();
     if (ret)
          return ret;

     proc_one_dir = proc_mkdir("one", NULL);

     return 0;
}

/******************************************************************************/

#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 0)
static void __exit deregister_devices(void)
{
     int i;

     unregister_chrdev(one_major, "one");

     for (i = 0; i < NUM_MINORS; i++) {
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 26)
          device_destroy(one_class, MKDEV(one_major, i));
#elif LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
          class_device_destroy(one_class, MKDEV(one_major, i));
#else
          class_simple_device_remove(MKDEV(one_major, i));
#endif
#endif

#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 18)
          devfs_remove("one/%d", i);
#endif
     }

#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 2)
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2, 6, 13)
     class_destroy(one_class);
#else
     class_simple_destroy(one_class);
#endif
#endif

#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 18)
     devfs_remove("one");
#endif
}
#else
static void __exit deregister_devices(void)
{
     int i;

     devfs_unregister_chrdev(one_major, "one");

     for (i = 0; i < NUM_MINORS; i++)
          devfs_unregister(devfs_handles[i]);
}
#endif

void __exit one_exit(void)
{
     deregister_devices();

     remove_proc_entry("one", NULL);

     one_core_exit( one_core );
}

module_init(one_init);
module_exit(one_exit);

